Build fails with "Method too large" exception

### Describe the bug

I have an application, which uses a lot (94) of different extensions. When I am trying to build it using the current SNAPSHOT, it fails with an exception during the build phase. If I use Quarkus 3.16.3, the apps works without any problems, same happens if I remove `quarkus-redis-client` extension

### Expected behavior

Application either compiles or fails with a more telling error

### Actual behavior

```
[error]: Build step io.quarkus.deployment.steps.ConfigGenerationBuildStep#generateConfigClass threw an exception: org.objectweb.asm.MethodTooLargeException: Method too large: io/quarkus/runtime/generated/Config.generateMappedProperties ()Ljava/util/Set;
[ERROR] 	at org.objectweb.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2088)
[ERROR] 	at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:506)
[ERROR] 	at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:246)
[ERROR] 	at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:257)
[ERROR] 	at io.quarkus.deployment.configuration.RunTimeConfigurationGenerator$GenerateOperation.run(RunTimeConfigurationGenerator.java:519)
[ERROR] 	at io.quarkus.deployment.steps.ConfigGenerationBuildStep.generateConfigClass(ConfigGenerationBuildStep.java:318)
[ERROR] 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:499)
```

### How to Reproduce?

1. `git clone git@github.com:quarkus-qe/beefy-scenarios.git && cd beefy-scenarios/002-quarkus-all-extensions`
2. `mvn clean verify # or mvn clean install -DskipTests`

### Output of `uname -a` or `ver`

6.11.7-300.fc41.x86_64

### Output of `java -version`

21.0.1, vendor: Eclipse Adoptium

### Quarkus version or git rev

26db59d0b672053df63fd5c058b0778ffaac4739

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java
@@ -182,7 +182,7 @@ public final class ExtensionLoader {
             }
 
             // ConfigMapping
-            ConfigClass mapping = readResult.getAllMappings().get(entry.getKey());
+            ConfigClass mapping = readResult.getAllMappingsByClass().get(entry.getKey());
             if (mapping != null) {
                 mappingClasses.put(entry.getValue(), mapping);
                 continue;

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java
@@ -628,9 +628,9 @@ public final class BuildTimeConfigurationReader {
                 objectsByClass.put(mapping.getKlass(), config.getConfigMapping(mapping.getKlass(), mapping.getPrefix()));
             }
 
-            Set<PropertyName> buildTimeNames = getMappingsNames(buildTimeMappings);
-            Set<PropertyName> buildTimeRunTimeNames = getMappingsNames(buildTimeRunTimeMappings);
-            Set<PropertyName> runTimeNames = getMappingsNames(runTimeMappings);
+            Set<PropertyName> buildTimeNames = mappingsToNames(buildTimeMappings).keySet();
+            Set<PropertyName> buildTimeRunTimeNames = mappingsToNames(buildTimeRunTimeMappings).keySet();
+            Set<PropertyName> runTimeNames = mappingsToNames(runTimeMappings).keySet();
             for (String property : allProperties) {
                 PropertyName name = new PropertyName(property);
                 if (buildTimeNames.contains(name)) {
@@ -1222,12 +1222,32 @@ public final class BuildTimeConfigurationReader {
             }
         }
 
-        private static Set<PropertyName> getMappingsNames(final List<ConfigClass> configMappings) {
+        private static Map<PropertyName, String> mappingsToNames(final List<ConfigClass> configMappings) {
             Set<String> names = new HashSet<>();
             for (ConfigClass configMapping : configMappings) {
                 names.addAll(ConfigMappings.getProperties(configMapping).keySet());
             }
-            return PropertiesUtil.toPropertyNames(names);
+            Map<PropertyName, String> propertyNames = new HashMap<>();
+            for (String name : names) {
+                PropertyName propertyName = new PropertyName(name);
+                if (propertyNames.containsKey(propertyName)) {
+                    String existing = propertyNames.remove(propertyName);
+                    if (existing.length() < name.length()) {
+                        propertyNames.put(new PropertyName(existing), existing);
+                    } else if (existing.length() > name.length()) {
+                        propertyNames.put(propertyName, name);
+                    } else {
+                        if (existing.indexOf('*') <= name.indexOf('*')) {
+                            propertyNames.put(new PropertyName(existing), existing);
+                        } else {
+                            propertyNames.put(propertyName, name);
+                        }
+                    }
+                } else {
+                    propertyNames.put(propertyName, name);
+                }
+            }
+            return propertyNames;
         }
     }
 
@@ -1249,7 +1269,9 @@ public final class BuildTimeConfigurationReader {
         final List<ConfigClass> buildTimeMappings;
         final List<ConfigClass> buildTimeRunTimeMappings;
         final List<ConfigClass> runTimeMappings;
-        final Map<Class<?>, ConfigClass> allMappings;
+        final List<ConfigClass> allMappings;
+        final Map<Class<?>, ConfigClass> allMappingsByClass;
+        final Map<PropertyName, String> allMappingsNames;
 
         final Set<String> unknownBuildProperties;
         final Set<String> deprecatedRuntimeProperties;
@@ -1273,7 +1295,9 @@ public final class BuildTimeConfigurationReader {
             this.buildTimeMappings = builder.getBuildTimeMappings();
             this.buildTimeRunTimeMappings = builder.getBuildTimeRunTimeMappings();
             this.runTimeMappings = builder.getRunTimeMappings();
-            this.allMappings = mappingsToMap(builder);
+            this.allMappings = new ArrayList<>(mappingsToMap(builder).values());
+            this.allMappingsByClass = mappingsToMap(builder);
+            this.allMappingsNames = ReadOperation.mappingsToNames(allMappings);
 
             this.unknownBuildProperties = builder.getUnknownBuildProperties();
             this.deprecatedRuntimeProperties = builder.deprecatedRuntimeProperties;
@@ -1355,8 +1379,16 @@ public final class BuildTimeConfigurationReader {
             return runTimeMappings;
         }
 
-        public Map<Class<?>, ConfigClass> getAllMappings() {
+        public List<ConfigClass> getAllMappings() {
             return allMappings;
+        }
+
+        public Map<Class<?>, ConfigClass> getAllMappingsByClass() {
+            return allMappingsByClass;
+        }
+
+        public Map<PropertyName, String> getAllMappingsNames() {
+            return allMappingsNames;
         }
 
         public Set<String> getUnknownBuildProperties() {

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java
@@ -6,6 +6,7 @@ import static io.quarkus.runtime.annotations.ConfigPhase.RUN_TIME;
 
 import java.lang.reflect.Field;
 import java.lang.reflect.Modifier;
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -63,6 +64,7 @@ import io.quarkus.runtime.configuration.HyphenateEnumConverter;
 import io.quarkus.runtime.configuration.NameIterator;
 import io.quarkus.runtime.configuration.PropertiesUtil;
 import io.quarkus.runtime.configuration.QuarkusConfigFactory;
+import io.smallrye.config.ConfigMappingInterface;
 import io.smallrye.config.ConfigMappings;
 import io.smallrye.config.ConfigMappings.ConfigClass;
 import io.smallrye.config.Converters;
@@ -88,9 +90,6 @@ public final class RunTimeConfigurationGenerator {
     public static final MethodDescriptor C_READ_CONFIG = MethodDescriptor.ofMethod(CONFIG_CLASS_NAME, "readConfig", void.class);
 
     static final FieldDescriptor C_MAPPED_PROPERTIES = FieldDescriptor.of(CONFIG_CLASS_NAME, "mappedProperties", Set.class);
-    static final MethodDescriptor C_GENERATE_MAPPED_PROPERTIES = MethodDescriptor.ofMethod(CONFIG_CLASS_NAME,
-            "generateMappedProperties", Set.class);
-    static final MethodDescriptor PN_NEW = MethodDescriptor.ofConstructor(PropertyName.class, String.class);
     static final FieldDescriptor C_UNKNOWN = FieldDescriptor.of(CONFIG_CLASS_NAME, "unknown", Set.class);
     static final FieldDescriptor C_UNKNOWN_RUNTIME = FieldDescriptor.of(CONFIG_CLASS_NAME, "unknownRuntime", Set.class);
 
@@ -192,9 +191,13 @@ public final class RunTimeConfigurationGenerator {
     static final MethodDescriptor PU_IS_PROPERTY_IN_ROOTS = MethodDescriptor.ofMethod(PropertiesUtil.class, "isPropertyInRoots",
             boolean.class, String.class, Set.class);
     static final MethodDescriptor HS_NEW = MethodDescriptor.ofConstructor(HashSet.class);
+    static final MethodDescriptor HS_NEW_SIZED = MethodDescriptor.ofConstructor(HashSet.class, int.class);
     static final MethodDescriptor HS_ADD = MethodDescriptor.ofMethod(HashSet.class, "add", boolean.class, Object.class);
+    static final MethodDescriptor HS_ADD_ALL = MethodDescriptor.ofMethod(HashSet.class, "addAll", boolean.class,
+            Collection.class);
     static final MethodDescriptor HS_CONTAINS = MethodDescriptor.ofMethod(HashSet.class, "contains", boolean.class,
             Object.class);
+    static final MethodDescriptor PN_NEW = MethodDescriptor.ofConstructor(PropertyName.class, String.class);
 
     // todo: more space-efficient sorted map impl
     static final MethodDescriptor TM_NEW = MethodDescriptor.ofConstructor(TreeMap.class);
@@ -262,7 +265,6 @@ public final class RunTimeConfigurationGenerator {
             roots = Assert.checkNotNullParam("builder.roots", builder.getBuildTimeReadResult().getAllRoots());
             additionalTypes = Assert.checkNotNullParam("additionalTypes", builder.getAdditionalTypes());
             cc = ClassCreator.builder().classOutput(classOutput).className(CONFIG_CLASS_NAME).setFinal(true).build();
-            generateMappedProperties();
             generateEmptyParsers();
             // not instantiable
             try (MethodCreator mc = cc.getMethodCreator(MethodDescriptor.ofConstructor(CONFIG_CLASS_NAME))) {
@@ -282,13 +284,16 @@ public final class RunTimeConfigurationGenerator {
             clinit.setModifiers(Opcodes.ACC_STATIC);
 
             cc.getFieldCreator(C_MAPPED_PROPERTIES).setModifiers(Opcodes.ACC_STATIC);
-            clinit.writeStaticField(C_MAPPED_PROPERTIES, clinit.invokeStaticMethod(C_GENERATE_MAPPED_PROPERTIES));
+            clinit.writeStaticField(C_MAPPED_PROPERTIES, clinit.newInstance(HS_NEW_SIZED,
+                    clinit.load((int) ((float) buildTimeConfigResult.getAllMappingsNames().size() / 0.75f + 1.0f))));
 
             cc.getFieldCreator(C_UNKNOWN).setModifiers(Opcodes.ACC_STATIC);
             clinit.writeStaticField(C_UNKNOWN, clinit.newInstance(HS_NEW));
 
             cc.getFieldCreator(C_UNKNOWN_RUNTIME).setModifiers(Opcodes.ACC_STATIC);
             clinit.writeStaticField(C_UNKNOWN_RUNTIME, clinit.newInstance(HS_NEW));
+
+            generateMappedProperties();
 
             clinitNameBuilder = clinit.newInstance(SB_NEW);
 
@@ -1191,26 +1196,32 @@ public final class RunTimeConfigurationGenerator {
         }
 
         private void generateMappedProperties() {
-            Set<String> names = new HashSet<>();
-            for (ConfigClass buildTimeMapping : buildTimeConfigResult.getBuildTimeMappings()) {
-                names.addAll(ConfigMappings.getProperties(buildTimeMapping).keySet());
-            }
-            for (ConfigClass staticConfigMapping : buildTimeConfigResult.getBuildTimeRunTimeMappings()) {
-                names.addAll(ConfigMappings.getProperties(staticConfigMapping).keySet());
-            }
-            for (ConfigClass runtimeConfigMapping : buildTimeConfigResult.getRunTimeMappings()) {
-                names.addAll(ConfigMappings.getProperties(runtimeConfigMapping).keySet());
-            }
-            Set<PropertyName> propertyNames = PropertiesUtil.toPropertyNames(names);
-
-            MethodCreator mc = cc.getMethodCreator(C_GENERATE_MAPPED_PROPERTIES);
+            MethodDescriptor method = MethodDescriptor.ofMethod(CONFIG_CLASS_NAME, "addMappedProperties", void.class);
+            MethodCreator mc = cc.getMethodCreator(method);
             mc.setModifiers(Opcodes.ACC_PRIVATE | Opcodes.ACC_STATIC);
-            ResultHandle set = mc.newInstance(HS_NEW);
-            for (PropertyName propertyName : propertyNames) {
-                mc.invokeVirtualMethod(HS_ADD, set, mc.newInstance(PN_NEW, mc.load(propertyName.getName())));
+            for (ConfigClass mapping : buildTimeConfigResult.getAllMappings()) {
+                mc.invokeStaticMethod(generateMappedProperties(mapping, buildTimeConfigResult.getAllMappingsNames()));
             }
-            mc.returnValue(set);
+            mc.returnVoid();
             mc.close();
+            clinit.invokeStaticMethod(method);
+        }
+
+        private MethodDescriptor generateMappedProperties(final ConfigClass mapping,
+                final Map<PropertyName, String> propertyNames) {
+            MethodDescriptor method = MethodDescriptor.ofMethod(CONFIG_CLASS_NAME,
+                    "addMappedProperties$" + mapping.getKlass().getName().replace('.', '$'), void.class);
+            MethodCreator mc = cc.getMethodCreator(method);
+            mc.setModifiers(Opcodes.ACC_PRIVATE | Opcodes.ACC_STATIC);
+            Map<String, ConfigMappingInterface.Property> properties = ConfigMappings.getProperties(mapping);
+            ResultHandle set = mc.readStaticField(C_MAPPED_PROPERTIES);
+            for (String propertyName : properties.keySet()) {
+                String name = propertyNames.get(new PropertyName(propertyName));
+                mc.invokeVirtualMethod(HS_ADD, set, mc.newInstance(PN_NEW, mc.load(name)));
+            }
+            mc.returnVoid();
+            mc.close();
+            return method;
         }
 
         private void reportUnknown(BytecodeCreator bc, ResultHandle unknownProperty) {

--- a/core/runtime/src/main/java/io/quarkus/runtime/configuration/PropertiesUtil.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/configuration/PropertiesUtil.java
@@ -1,10 +1,6 @@
 package io.quarkus.runtime.configuration;
 
-import java.util.HashMap;
-import java.util.Map;
 import java.util.Set;
-
-import io.smallrye.config.PropertyName;
 
 public class PropertiesUtil {
     private PropertiesUtil() {
@@ -47,29 +43,5 @@ public class PropertiesUtil {
 
     public static boolean isPropertyQuarkusCompoundName(NameIterator propertyName) {
         return propertyName.getName().startsWith("\"quarkus.");
-    }
-
-    public static Set<PropertyName> toPropertyNames(final Set<String> names) {
-        Map<PropertyName, String> propertyNames = new HashMap<>();
-        for (String name : names) {
-            PropertyName propertyName = new PropertyName(name);
-            if (propertyNames.containsKey(propertyName)) {
-                String existing = propertyNames.remove(propertyName);
-                if (existing.length() < name.length()) {
-                    propertyNames.put(new PropertyName(existing), existing);
-                } else if (existing.length() > name.length()) {
-                    propertyNames.put(propertyName, name);
-                } else {
-                    if (existing.indexOf('*') <= name.indexOf('*')) {
-                        propertyNames.put(new PropertyName(existing), existing);
-                    } else {
-                        propertyNames.put(propertyName, name);
-                    }
-                }
-            } else {
-                propertyNames.put(propertyName, name);
-            }
-        }
-        return propertyNames.keySet();
     }
 }

